File: src/test/java/kitchenpos/application/ProductServiceTest.java
package kitchenpos.application;

import static java.math.BigDecimal.valueOf;
import kitchenpos.domain.MenuRepository;
import kitchenpos.domain.Product;
import kitchenpos.domain.ProductRepository;
import static kitchenpos.fixtures.ProductFixtures.양념치킨;
import static kitchenpos.fixtures.ProductFixtures.후라이드치킨;
import static kitchenpos.fixtures.ProductFixtures.후라이드치킨_가격;
import kitchenpos.infra.PurgomalumClient;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.UUID;

class ProductServiceTest {

    private ProductService productService;
    private ProductRepository productRepository;
    private MenuRepository menuRepository;
    private PurgomalumClient purgomalumClient;

    @BeforeEach
    void setUp() {
        productRepository = new InMemoryProductRepository();
        menuRepository = new InMemoryMenuRepository();
        purgomalumClient = new FakePurgomalumClient();
        productService = new ProductService(productRepository, menuRepository, purgomalumClient);
    }

    @Test
    void 상품을_등록할_수_있다() {
        // given
        Product request = 후라이드치킨();

        // when
        Product savedProduct = productService.create(request);

        // then
        assertThat(savedProduct).isNotNull();
        assertThat(savedProduct.getId()).isNotNull();
        assertThat(savedProduct.getName()).isEqualTo("후라이드치킨");
        assertThat(savedProduct.getPrice()).isEqualTo(valueOf(16000));
    }

    @Test
    void 상품_가격을_입력하지_않으면_등록할_수_없다() {
        // given
        Product request = new Product();
        request.setName(후라이드치킨().getName());
        request.setPrice(null);

        // when & then
        assertThatThrownBy(() -> productService.create(request))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void 상품_가격이_0원_미만이면_등록할_수_없다() {
        // given
        Product request = new Product();
        request.setName(후라이드치킨().getName());
        request.setPrice(valueOf(-16000));

        // when & then
        assertThatThrownBy(() -> productService.create(request))
            .isInstanceOf(IllegalArgumentException.class);
    }


    @Test
    void 상품_이름이_존재하지_않으면_등록할_수_없다() {
        // given
        Product request = new Product();
        request.setName(null);
        request.setPrice(후라이드치킨_가격);

        // when & then
        assertThatThrownBy(() -> productService.create(request))
            .isInstanceOf(IllegalArgumentException.class);
    }
    @Test
    void 상품의_가격을_변경할_수_있다() {
        // given
        Product request = 후라이드치킨();
        productRepository.save(request);

        UUID productId = request.getId();
        BigDecimal 후라이드치킨_가격 = BigDecimal.valueOf(16000);

        // when
        Product updateProduct = new Product();
        updateProduct.setPrice(후라이드치킨_가격);
        productService.changePrice(productId, updateProduct);

        // then
        Product updatedProduct = productRepository.findById(productId).orElseThrow();
        // 값만 비교
        assertThat(updatedProduct.getPrice().compareTo(후라이드치킨_가격)).isEqualTo(0);

        // scale 제거하고 비교
        assertThat(updatedProduct.getPrice().stripTrailingZeros())
            .isEqualTo(후라이드치킨_가격.stripTrailingZeros()); // stripTrailingZeros()를 사용하면 소수점이 필요 없는 경우 자동으로 정리함
    }

    @Test
    void 상품_가격_변경시_가격이_null_이면_변경할_수_없다() {
        // given
        Product request = 후라이드치킨();
        productRepository.save(request);

        UUID productId = request.getId();

        // when
        Product updateProduct = new Product();
        updateProduct.setPrice(null);

        // then
        assertThatThrownBy(() -> productService.changePrice(productId, updateProduct))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void 상품_가격_변경시_가격이_0원_미만이면_변경할_수_없다() {
        // given
        Product request = 후라이드치킨();
        productRepository.save(request);

        UUID productId = request.getId();
        BigDecimal 후라이드치킨_가격 = BigDecimal.valueOf(-16000);

        // when
        Product updateProduct = new Product();
        updateProduct.setPrice(후라이드치킨_가격);

        // then
        assertThatThrownBy(() -> productService.changePrice(productId, updateProduct))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void 존재하지_않은_상품_ID로_가격을_변경할_수_없다() {
        // given
        UUID nonExistentProductUd = UUID.randomUUID();
        BigDecimal changedPrice = BigDecimal.valueOf(16000);

        Product updateProduct = new Product();
        updateProduct.setPrice(changedPrice);

        // when & then
        assertThatThrownBy(() -> productService.changePrice(nonExistentProductUd, updateProduct))
            .isInstanceOf(NoSuchElementException.class);
    }

    @Test
    void 등록된_상품을_전체_조회할_수_있다() {
        // given
        Product burger = 후라이드치킨();
        Product pizza = 양념치킨();

        // when
        productService.create(burger);
        productService.create(pizza);

        // then
        List<Product> products = productRepository.findAll();
        assertNotNull(products);
        assertThat(products).hasSize(2);
    }
}

File: src/test/java/kitchenpos/application/ProductServiceTest.kt
package kitchenpos.application

import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import kitchenpos.domain.Menu
import kitchenpos.domain.MenuProduct
import kitchenpos.domain.MenuRepository
import kitchenpos.domain.ProductRepository
import kitchenpos.infra.PurgomalumClient
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import random.randomProduct
import random.randomProductId
import java.math.BigDecimal
import java.util.*

class ProductServiceTest {
    private val productRepository: ProductRepository = mockk()
    private val menuRepository: MenuRepository = mockk()
    private val purgomalumClient: PurgomalumClient = mockk()
    private lateinit var productService: ProductService

    @BeforeEach
    fun setUp() {
        productService = ProductService(productRepository, menuRepository, purgomalumClient)
    }

    @Nested
    inner class `상품 생성시` {
        @Test
        fun `가격이 널인 경우 예외 발생`() {
            // given
            val product = randomProduct(price = null)
            // when then
            assertThrows<IllegalArgumentException> {
                productService.create(product)
            }
        }

        @Test
        fun `가격이 음수인 경우 예외 발생`() {
            // given
            val product = randomProduct(price = BigDecimal.valueOf(-1))
            // when then
            assertThrows<IllegalArgumentException> {
                productService.create(product)
            }
        }

        @Test
        fun `상품명이 널인 경우 예외 발생`() {
            // given
            val product = randomProduct(name = null)
            // when then
            assertThrows<IllegalArgumentException> {
                productService.create(product)
            }
        }

        @Test
        fun `상품명이 비속어가 포함될 경우 예외 발생`() {
            // given
            val product = randomProduct()
            every { purgomalumClient.containsProfanity(product.name) } returns true

            // when then
            assertThrows<IllegalArgumentException> {
                productService.create(product)
            }
        }

        @Test
        fun `정상 생성`() {
            // given
            val product = randomProduct()
            every { purgomalumClient.containsProfanity(product.name) } returns false
            every { productRepository.save(any()) } returns product
            // when
            val expected = productService.create(product)

            // then
            expected shouldBe product
        }
    }

    @Nested
    inner class `상품 가격 변경시` {
        @Test
        fun `변경 요청 가격이 널인 경우 예외 발생`() {
            // given
            val product = randomProduct(price = null)
            val productId = randomProductId()

            // when then
            assertThrows<IllegalArgumentException> {
                productService.changePrice(productId, product)
            }
        }

        @Test
        fun `변경 요청 가격이 음수인 경우 예외 발생`() {
            // given
            val product = randomProduct(price = BigDecimal.valueOf(-1))
            val productId = randomProductId()

            // when then
            assertThrows<IllegalArgumentException> {
                productService.changePrice(productId, product)
            }
        }

        @Test
        fun `존재하지 않는 상품인 경우 예외 발생`() {
            // given
            val product = randomProduct()
            val productId = randomProductId()
            every { productRepository.findById(productId) } returns Optional.empty()
            // when then
            assertThrows<NoSuchElementException> {
                productService.changePrice(productId, product)
            }
        }

        @Test
        fun `메뉴 가격이 상품의 합산 가격보다 클 경우, 해당 메뉴 상품을 비노출 처리`() {
            // given
            val productId = UUID.randomUUID()
            val price = BigDecimal.valueOf(200)
            val requestProduct = randomProduct(id = productId, price = price)
            every { productRepository.findById(productId) } returns Optional.of(randomProduct(id = productId, price = BigDecimal.valueOf(100)))

            val menu = mockk<Menu>(relaxed = true)
            val menuProduct = mockk<MenuProduct>()
            val menuProducts = listOf(menuProduct)
            every { menuRepository.findAllByProductId(productId) } returns listOf(menu)

            every { menuProduct.product } returns requestProduct
            every { menuProduct.quantity } returns 1
            every { menu.price } returns BigDecimal.valueOf(250) // Menu price greater than product total price
            every { menu.menuProducts } returns menuProducts

            // when
            productService.changePrice(productId, requestProduct)

            // then
            verify { menu.setDisplayed(false) }
        }
    }

    @Test
    fun `상품 전체 조회`() {
        // given
        val products = listOf(randomProduct(), randomProduct(), randomProduct())
        every { productRepository.findAll() } returns products

        // when then
        productService.findAll() shouldBe products
    }
}
